Support poll vote/unvote

PIG208 opened the pull request:
~~This is stacked on top of #912.~~

Fixes: #166

PIG208 commented:
<details>
<summary>Preview</summary>

![1000014790](https://github.com/user-attachments/assets/00c5f9cc-54d3-438f-9891-cebf5e2c66d2)
![1000014789](https://github.com/user-attachments/assets/ed70b05c-869c-447c-b7a9-d696003d61ea)


</details>

PIG208 commented:
<details>

<summary>Preview</summary>
| | |
| - | - |
| ![Screenshot from 2024-10-07 19-55-57](https://github.com/user-attachments/assets/4e22364c-95c5-48bd-b5e3-537d3341a41c) | ![Screenshot from 2024-10-07 19-57-28](https://github.com/user-attachments/assets/42530117-5f24-4656-8153-2461a69a1259) |

</details>

chrisbobbe commented on lib/api/route/submessage.dart:
```diff
@@ -0,0 +1,19 @@
+    // There is also 'zform' for trivia quiz bot, but we currently do not have
+    // plan to support that.
+    'msg_type': RawParameter('widget'),
+    'content': RawParameter(jsonEncode(content)),
```
Would we get the same result from `'content': content` as we do from this, and if so, would it be preferable because it looks simpler?

PIG208 replied:
The case of `content` is a bit peculiar, because the API expects a plain string instead of an object, while the string is expected to be some JSON encoded value.

I think it would be good to mention this in comments.

chrisbobbe commented on test/model/message_test.dart:
```diff
@@ -628,13 +628,13 @@ void main() {
-              options: (options != null)
-                ? options.map((e) => e.text).toList()
+              options: (optionVoterIdsPairs != null)
+                ? optionVoterIdsPairs.map((e) => e.$1).toList()
```
Oh, odd, I guess `$1` refers to the first element of the tuple? So it's not zero-indexed.

chrisbobbe commented on lib/widgets/poll.dart:
```diff
@@ -68,15 +68,14 @@ class _PollWidgetState extends State<PollWidget> {
-          children: [
-            ConstrainedBox(
-              constraints: const BoxConstraints(minWidth: 25).tighten(height: 25),
+      return Row(
```
```
poll [nfc]: Move vote count paddings to a box
```

> This does not visually change the paddings.

This doesn't seem to be true. Here's before/after this change, with message https://chat.zulip.org#narrow/stream/7-test-here/topic/new.20poll.20test/near/1956979 :

| Before | After |
| --- | --- |
| <img width="559" alt="image" src="https://github.com/user-attachments/assets/0d7bd6d8-f7ac-4aaf-b0be-eff246266a97"> | <img width="559" alt="image" src="https://github.com/user-attachments/assets/ab890e47-d01b-41a7-802c-ed75d28b7e8b"> |

When I open the two images in different browser tabs and switch between them, I see a layout change. Do you reproduce?

PIG208 replied:
Hmm, I can see that in the screenshots, but not on my Android device.

chrisbobbe replied:
When the `bottom: 5` padding wraps the whole row (before this commit), instead of just the vote-count rectangle, there will be 5px whitespace below the text (option name + voters' names) even when the text takes more vertical space than the vote-count rectangle.

I think you can reproduce if you make an option with a name so long that it takes two or three lines on your device.

chrisbobbe commented on lib/widgets/poll.dart:
```diff
@@ -73,10 +74,14 @@ class _PollWidgetState extends State<PollWidget> {
+                end: 5, top: verticalPadding, bottom: verticalPadding),
               child: Container(
+                // This is only in effect
+                // when the vote count has more than 2 digits.
```
I'm a little suspicious of a comment like this :) because I don't know if it's true as written for all values of the user's system bold-text and text-size settings. If this needs a comment, maybe it could be something like:

```dart
                  // Inner padding preserves whitespace even when the text's
                  // width approaches the button's min-width (e.g. because 
                  // there are three digits).
```

chrisbobbe reviewed (commented):
Thanks! Comments below.

Also, I noticed that the web app has a "pressed" style for the button. Do we want to either do that, or make an issue and TODO for it?

I guess my cursor isn't captured in this screenshot :) but you can imagine it there:

![image](https://github.com/user-attachments/assets/8d25fad7-cb65-4896-a233-2ee9749b01a0)

alya commented:
Screenshots and CZO discussion look good to me!

PIG208 commented:
The "pressed" UX might fit better in a separate PR. I guess we need to have a way to mark some data as "dirty" until an event comes and update it (but there could also be cases where we expect multiple events).

I also noticed that my previous screenshots have a non-default sized text. The 22px font size for the vote count text is actually inadequate. Updated that to 20px with new screenshots:

<details>
<summary>Updated screenshots</summary>


| | | |
| - | - | - |
| ![1000014904](https://github.com/user-attachments/assets/d39ce767-4260-42ab-bde1-e7b39fbf8ad3) | ![1000014902](https://github.com/user-attachments/assets/ce9538ed-2f01-47dc-b454-e454c09ac692) | ![1000014901](https://github.com/user-attachments/assets/61a44e5e-3c82-4d90-af3c-024d6af4c2f5) |


</details>

chrisbobbe commented:
> The "pressed" UX might fit better in a separate PR. I guess we need to have a way to mark some data as "dirty" until an event comes and update it (but there could also be cases where we expect multiple events).

I was just thinking about highlighting the button when the user's finger is on it; I wasn't thinking about the state between the tap and the event arriving.

gnprice commented:
> I think it might be the more complicated thing where it highlights the button between the tap and the event arrival.

That more complicated thing sounds like when it has an effect, it'd be liable to make it feel like the button was stuck. (Though I haven't tried to play with this aspect of web's UX.)

I'd be glad to have immediate feedback while the user's finger is there, and perhaps animating back to the resting state over a short fixed duration thereafter like 100ms, while not trying to have any longer-lived "dirty" state. This would make it similar to the existing behavior on the mark-as-read button.

chrisbobbe commented:
> That more complicated thing sounds like when it has an effect, it'd be liable to make it feel like the button was stuck. (Though I haven't tried to play with this aspect of web's UX.)

I agree; let's not do this. I think Vlad likes a scale animation for the "pressed" state, like we did for the mark-as-read button in 791b703c.

Then I think https://github.com/zulip/zulip-flutter/pull/939#discussion_r1794340316 is still pending, about a layout-change commit that should be unmarked NFC. One consequence of the change is that it removes any spacing between two options when the first one's text runs onto multiple lines:

| Before | After |
| --- | --- |
| <img width="685" alt="image" src="https://github.com/user-attachments/assets/ae7d06ba-1d7b-44e7-972a-cd72e3774ed6"> | <img width="685" alt="image" src="https://github.com/user-attachments/assets/62b03c4c-e800-4c50-b8f6-0ddfaf4134bb"> |

I don't love that consequence, but it's helped in a later commit that puts 2.5px spacing above the button.

chrisbobbe commented:
Then also, as discussed in the office just now: in the place where we use a fixed height, we'll want a min-height instead, so that it doesn't break with a large system text size setting:

| Current revision | Changed to min-height |
| --- | --- |
| <img width="685" alt="image" src="https://github.com/user-attachments/assets/1ff9dbb7-8581-4979-9063-f980e7bd1b3b"> | <img width="685" alt="image" src="https://github.com/user-attachments/assets/4e37f6eb-5b0b-4faa-ab7e-006671949f56"> |

PIG208 commented:
Updated the PR. The button feedback will be done in a separate PR. The current plan is to extract a common animated button widget.

This should handle larger font sizes now, as well as consistent spacing; the padding box commit should be truly NFC now.

chrisbobbe commented:
Great! LGTM, marking for Greg's review.

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +86,70 @@ class _PollWidgetState extends State<PollWidget> {
+                // than the vote count box.  Still, we avoid padding at the
+                // start because we want to align all the poll options to the
+                // surrounding messages.
+                padding: const EdgeInsetsDirectional.only(
+                  end: 5, top: verticalPadding, bottom: verticalPadding),
```
nit:

    poll [nfc]: Move vote count paddings to a box

I'm not sure what "to a box" means — `Padding` is always a box, right?

Could say:

> poll [nfc]: Move vote count paddings inside children of Row

and I think that accurately conveys what this commit is doing.

gnprice replied:
other nit in same commit message:

    specified on the outer (now removed) Padding.  The Container is
    expanded because it now also contains the paddings.

I think you mean the ConstrainedBox, not Container.

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +86,70 @@ class _PollWidgetState extends State<PollWidget> {
+            child: Padding(
+              // When the bottom of the text reaches farther than the vote count
+              // box's padded bottom edge, this padding helps ensure that we
+              // still maintain a consistent spacing of `verticalPadding * 2`
+              // logical pixels between option rows.
+              padding: const EdgeInsets.only(bottom: verticalPadding),
```
The comment and code don't seem to agree here. Do you want `EdgeInsets.symmetric(vertical: verticalPadding)`?

PIG208 replied:
The other `verticalPadding` comes from the top of the next vote count box (in the next row); combined we get `verticalPadding * 2`. This was added to fix https://github.com/zulip/zulip-flutter/pull/939#issuecomment-2406034465, but I think the comment can articulate that purpose better.

gnprice replied:
(Thread continues below: https://github.com/zulip/zulip-flutter/pull/939#discussion_r1815952172 )

gnprice commented on lib/api/model/submessage.dart:
```diff
@@ -404,6 +404,9 @@ class Poll extends ChangeNotifier {
   final Map<PollOptionKey, PollOption> _options = {};
 
+  bool hasUserVotedFor({required int userId, required PollOptionKey key}) =>
+    _options.containsKey(key) && _options[key]!.voters.contains(userId);
```
```suggestion
    _options[key]?.voters.contains(userId) ?? false;
```

That avoids doing the lookup twice.

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +86,70 @@ class _PollWidgetState extends State<PollWidget> {
+        textBaseline: localizedTextBaseline(context),
+        children: [
+          GestureDetector(
+            // TODO: Implement feedback when the user taps the button
+            onTap: () => _toggleVote(option),
```
nit: squash the TODO-comment commit into the commit that adds the code it's commenting on

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -44,21 +48,34 @@ class _PollWidgetState extends State<PollWidget> {
 
+  void _toggleVote(PollOption option) async {
+    final store = PerAccountStoreWidget.of(context);
+    // The poll data in store might be obselete before we get the event
+    // that updates it.  This is fine because the result will be consistent
+    // eventually, regardless of the possible duplicate requests.
+    final op = widget.poll.hasUserVotedFor(userId: store.selfUserId, key: option.key)
+      ? PollVoteOp.remove
```
I don't quite follow what this comment is saying. What's the scenario where it's obsolete, and when?

If this is intended as reasoning for why we're determining `op` based on the data in the store, rather than something else like telling the server to toggle from whatever state the server has:

I think the real point there is that when the UI shows the user a state X and the user says to change it to state Y, the thing we want to tell the server is to please change in the direction X → Y. We basically never want to translate that into something like "please flip between state X or Y, whichever one you currently have it in". (The only exception would be if the nature of "state X" and "state Y" is somehow such that that translation would be a natural interpretation of what the user actually intends.)

So, here, the UI we're showing either has the user as having already voted for the option, or as not having done so. If the UI says the user has already voted and the user hits the button, that means to remove their vote, unconditionally. Even if the server currently thinks they haven't voted, what the user asked for is to remove their vote, not add it — after all, they were looking at a UI that said they had voted.

One consequence of that general principle is that the canonical thing to do here wouldn't be to go look up `option.key` in the store again and find what it says. I think in this revision that will end up having the same behavior we want… but the canonical way to get that behavior is to look directly at the data structure that was used for rendering the UI. Here, `option.voters` is what controlled the list of names we showed. So how about we just inspect `option.voters` directly?

PIG208 replied:
> I think the real point there is that when the UI shows the user a state X and the user says to change it to state Y, the thing we want to tell the server is to please change in the direction X → Y. We basically never want to translate that into something like "please flip between state X or Y, whichever one you currently have it in". 

Yeah exactly that. The user couldn't instantly toggle the vote because it takes time to send an update to the server and then receive the event for it — that's not really a bad thing.

In the legacy mobile app, with locally echoing, you can have a sequence of states like this:
- User taps the button to unvote;
- User taps the button again to vote;
- The submessage event for the unvote arrives;
- The submessage event for the vote arrives;

which result in visually four toggles. This is probably because we don't have a way to deduplicate the updates. The current behavior of the this PR revision doesn't have that problem.

gnprice reviewed (commented):
Thanks @PIG208 for building this, and @chrisbobbe for the previous reviews!

Generally this looks good. Comments below.

PIG208 commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +86,70 @@ class _PollWidgetState extends State<PollWidget> {
+                    color: theme.colorPollVoteCountBackground,
+                    border: Border.all(color: theme.colorPollVoteCountBorder),
+                    borderRadius: BorderRadius.circular(3)),
+                  child: Center(
+                    child: Text(option.voters.length.toString(),
+                      textAlign: TextAlign.center,
```
Tested that I could see none. I checked my reflog and found that they have coexisted in some earliest draft of the read-only `PollWidget`. It's likely that they were just redundant before and lived.

PIG208 commented:
This PR has been updated. Thanks for the review!

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +84,62 @@ class _PollWidgetState extends State<PollWidget> {
           store.users[userId]?.fullName ?? zulipLocalizations.unknownUserName)
         .join(', ');
 
-      return Padding(
-        padding: const EdgeInsets.only(bottom: 5),
```
poll: Move vote count paddings inside children of Row

It looks like in this revision, this commit actually moves the padding only inside the first child, and drops the padding from the second one.

    This preserves most of the visuals. Except that when the vote count
    box is shorter than the voter names text, the padding between rows
    is gone.

That seems undesirable. What if we include padding on both children, like in the previous revision?

I think this change in this revision may have been in response to this thread:
https://github.com/zulip/zulip-flutter/pull/939#discussion_r1809681899
but I don't really understand the motivation for removing the padding from the voter names.

gnprice reviewed (commented):
Thanks for the revision! Generally this looks good; one question below.

Since this revision changed the layout, please also post updated screenshots.

gnprice reviewed (commented):
Great, thanks for changing that back.

Small comments below — most of them about a follow-up in that same area.

PIG208 commented on lib/widgets/poll.dart:
```diff
@@ -69,46 +84,68 @@ class _PollWidgetState extends State<PollWidget> {
+        Padding(
+          // `verticalPadding` out of 6 logical pixels come from the first
+          // option row.
+          padding: const EdgeInsets.only(bottom: 6 - verticalPadding),
```
> Ah, I guess if the text is long, then there's still top padding because CrossAxisAlignment.baseline means the option text starts at the same level as the count and expands down from there, and so the button still extends further upward from the baseline than the text does.

Yeah, exactly. I've made them symmetric in the new revision.

PIG208 commented:
Pushed a new revision. Thanks for the review!

- Now we use vertically symmetric padding for all the option items, and limit the one-sided paddings.
- This also fixed an issue with the "This poll has no options yet" text where the bottom padding is missing.
- I added `5 - verticalPadding` to the entire poll widget so that there is still 5px space after it.

| before | after |
| - | - |
| ![1000015159](https://github.com/user-attachments/assets/8e5d4d3f-4e6a-40c6-9188-92575b9d45fb) | ![1000015158](https://github.com/user-attachments/assets/b9b901c5-aeb9-42b8-a4e9-2e1f1fa041a4) |

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,48 +84,74 @@ class _PollWidgetState extends State<PollWidget> {
+            padding: const EdgeInsets.only(bottom: 6 - verticalPadding),
+            child: question),
+          if (widget.poll.options.isEmpty)
+            Padding(
+              // This is consistent with the option rows' padding.
```
This change makes this commit no longer NFC, right?
ceca4cec7 poll [nfc]: Move vote count paddings inside children of Row

It's also not part of what that summary describes. So I think it should just become a separate commit, fixing that layout bug you noted.

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,48 +84,74 @@ class _PollWidgetState extends State<PollWidget> {
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          Padding(
+            // `verticalPadding` out of 6 logical pixels
+            // come from the first option row.
+            padding: const EdgeInsets.only(bottom: 6 - verticalPadding),
+            child: question),
```
nit: this is a bit simpler to express with a SizedBox spacer:
```suggestion
          question,
          // `verticalPadding` out of 6px come from the first option row.
          const SizedBox(height: 6 - verticalPadding),
```

gnprice commented on lib/widgets/poll.dart:
```diff
@@ -69,48 +84,74 @@ class _PollWidgetState extends State<PollWidget> {
-        for (final option in widget.poll.options)
-          buildOptionItem(option),
-      ]);
+    return Padding(
+      // `verticalPadding` out of 5 logical pixels
+      // come from the last option row.
+      padding: const EdgeInsets.only(bottom: 5 - verticalPadding),
```
Then similarly this padding can be expressed as a spacer at the end of the column's children. I think that makes it significantly clearer what's going on, because it keeps things in a consistent order.

gnprice reviewed (commented):
Thanks for the revision! Glad to have that layout issue fixed — and I think the layout logic all makes sense to me now. Just a few style comments below.

PIG208 commented:
Thanks! The PR has been updated.

gnprice commented:
Thanks! Looks good now — merging, with two tweaks:

    -    poll: Enlarge and vetically center vote count box
    +    poll: Enlarge and vertically center vote count box


    -+        // `verticalPadding` out of 6px come from the first option row.
    ++        // `verticalPadding` out of 6px comes from the first option row.

(I know the latter comes from my suggestion above; rereading, I'd incompletely edited it from the previous version.)
